ENHANCED: added API for blobs

kamahen opened the pull request.

kamahen committed:
ENHANCED: added API for blobs

kamahen committed:
Improve compare_fields() design to avoid cast_blob() calls

kamahen commented on SWI-cpp2.h:
```diff
@@ -1515,4 +1516,144 @@ template <typename ContextType> class PlForeignContextPtr
+    // TODO: following should write ")\n" if PL_WRT_NEWLINE but there
+    //       appears to be a bug in the top-level code that sets the
+    //       flag:
+    Sfprintf(s, PL_WRT_NEWLINE ? ")" : ")");
```
PL_WRT_NEWLINE has very terse documentation. Does it mean that the write callback is being called in the context of a top-level display of results? - because it doesn't seem to mean "put a newline after this", both by experimentation and by reading the code.

If so, I'll change this code (and also similar code in library(pcre) and possibly elsewhere).

JanWielemaker replied:
The flag is passed, but doesn't expect any action from the blob handler.    It does cause PL_write_term() to add a newline at the end.   This is convenient as it allows output with a newline in a single call.   Using a single call, no other output (from threads) can be between, so it also avoids having to lock the stream explicitly.

Most of the flags are meaningless.   The writer hook may decide to act on quoted write.

kamahen replied:
So, if I call PL_write_term() from inside the blob's "write" callback, I should use `flags & (~PL_WRT_NEWLINE)`? That would propagate, e.g. `PL_WRT_QUOTED`.

I've changed the code from
```cpp
    if ( pathname.not_null() )
      if ( !Sfprintf(s, ",path=%Ws", pathname.as_wstring().c_str()) )
         return false;
```
to
```cpp
    flags &= ~PL_WRT_NEWLINE;
    if ( pathname.not_null() )
      if ( !Sfprintf(s, ",path=") ||
           !PlTerm_atom(pathname).write(s, 1200, flags) ) // calls PL_term_write()
         return false;
``

JanWielemaker replied:
I would typically not call Prolog write anyway.   Surely I've never needed it.   In this case, I'd get the text from the atom as UTF-8 and use `%Us` with Sfprintf().  If you really want, I guess you can.   Only the `release` hook does not allow for callbacks as it will get atom-gc into a deadlock.    If you do a write callback you probably do not want to pass on any flags, possibly except a few specific ones.

kamahen committed:
Improve blob interface

kamahen committed:
adjust the write flag

kamahen committed:
Add convenience macro for defining blobs

kamahen committed:
Changed PlBlob to not use a template
  - the template seemed to confuse the compiler
Renamed my_data to dbref in the documentation and test code
  - same names as in rocksdb, which is being used validate the API

kamahen committed:
Put blob "vtable" functions into a class

JanWielemaker commented:
This pull request has been mentioned on **SWI-Prolog**. There might be relevant details there:

https://swi-prolog.discourse.group/t/changes-to-swi-cpp-h/5601/66

kamahen committed:
Add PlExceptionBase, for more convenient "atch" statements

JanWielemaker commented:
Thanks.  Merged after (partial) squash and fixing LaTeX errors
